src/core.ts

```diff
@@ -84,21 +84,15 @@ export function addEvent(event: IEventData, scheduleUpload: boolean = true) {
     type: event.type,
     state: event.state
   };
-  let eventStr = JSON.stringify(evt);
-  if (nextPayloadLength > 0 && nextPayloadLength + eventStr.length > config.batchLimit) {
-    uploadNextPayload();
-  }
-  nextPayload.push(eventStr);
-  nextPayloadLength += eventStr.length;
-
-  // Edge case:
-  // Don't schedule next upload when next payload consists of exactly one XhrError instrumentation event.
-  // This helps us avoid the infinite loop in the case when all requests fail (e.g. dropped internet connection)
-  // Infinite loop comes from sending instrumentation about failing to deliver previous delivery failure instrumentation.
-  let payloadIsSingleXhrErrorEvent = event.state && event.state.type === Instrumentation.XhrError && nextPayload.length === 1;
-  if (scheduleUpload && !payloadIsSingleXhrErrorEvent) {
+  let addEventMessage: IAddEventMessage = {
+    type: WorkerMessageType.AddEvent,
+    event: evt,
+    time: getTimestamp()
+  };
+  compressionWorker.postMessage(JSON.stringify(addEventMessage));
+  if (scheduleUpload) {
```

Comment:
>scheduleUpload [](start = 6, length = 14)

what do you think about passing scheduleUpload to the worker and allowing the worker to handle setting these timeouts instead of managing it at this layer? In case setting and clearing timeouts ever starts to take a lot of time again we can make sure it is on the background thread and it seems like decisions of when to call forceUpload can stay within the worker isntead of us setting a timer to then send a message to the worker to forceUpload

Reply:
I thought about it at some point. I think my main hunch against it was design practice. I didn't think it'd be good for a worker to do something independently and be "smart" about it. I wanted to maintain the control of data flow within core and make worker act strictly based on the command messages.

---
In reply to: [136670940](https://github.com/Microsoft/clarity-js/pull/28#discussion_r136670940) [](ancestors = 136670940)